### ISSUE TITLE
Perfusion model

### ISSUE THREAD

**Issue by mathieuboudreau** [qMRLab/qMRLab#315]
Just noticed @tanguyduval's working on this again, thought I'd open an issue so that everyone is aware. @tanguyduval feel free to provide any details/questions/discussion here.

branch: [Perfusion](https://github.com/qMRLab/qMRLab/tree/Perfusion)

**Comment by mathieuboudreau** [comment]
@ileppe were you also working on this, or planning on it? I've got a faint memory that someone mentioned it recently, but I can't put my finger on who or when.

**Comment by tanguyduval** [comment]
@mathieuboudreau I have DCE data so I wrote this module.

Currently, I'm assuming that the bolus input function is a dirac, but I have some code to (i) detect arteries and (2) deconvoluate the bolus the Arterial input Function measured in the arteries.
After that, there is significant work to implement the models of the bolus that are used to extract information such as cerebral blood flow, transi-time, ktrans, etc...

I think @notZaki mentionned about DCE. If you do have some code/models that you want to implement in this module, feel free to take the lead on this branch (I am not activelly developping it).

**Comment by ileppe** [comment]
@mathieuboudreau I also have some older code for many different fitting models, but I had sent whatever I had to Ives so it's very likely @notZaki has expanded way beyond those by now!

**Comment by notZaki** [comment]
@tanguyduval Is the branch also supposed to handle dynamic susceptibility contrast (DSC) data? The code seems to be using R2star which is typically used for DSC.

@mathieuboudreau There is a fair bit of pre-processing for DCE modelling---possible motion correction, T1 mapping, signal to concentration conversion, selection of arterial input function or reference region---some of which might require user interactivity. Should these steps also be included? This might be related to previous discussion on ASL #207 

@ileppe Thanks for sharing that code by the way. It was really helpful at getting me started.

**Comment by mathieuboudreau** [comment]
>@mathieuboudreau There is a fair bit of pre-processing for DCE modelling---possible motion correction, T1 mapping, signal to concentration conversion, selection of arterial input function or reference region---some of which might require user interactivity. Should these steps also be included? This might be related to previous discussion on ASL #207

Heh, this topic (if to include pre-processing or not, and if to write our own functions for them or call external tools like SPM) has been going on and off for the better part of a year. I don't think I have an answer for you at the moment; we have qMRLab meetings (located at the Poly campus, or via videocall for people not in town like me) every three weeks or so, so if you want you're certainly invited to join at an upcoming meeting to expand on what would be needed and what you've been doing so far in terms of pre-processing.

What I've been recommending has been to always try implement the easiest workflow, and work yourself up from there. i.e. first create a qMRLab model that assumes you've already had all your pre-processing done already (and/or anything absolutely essential/unique to your method), and wait to implement the pre-processing steps later one at time. This way, it gives the option for the user to choose the external pre-processing tools that they are comfortable with already, but leaves the door open for the development of the pre-processing tools when developers get the time later on. For example, we first developed the B1 mapping models, and now @ileppe has been working on implementing an image filtering step which is a common B1 mapping preprocessing step.

tl;dr do the largest/easier chunk first, if possible. But I don't have experience with DCE/ASL, so you're input on what's an essential pre-processing step to include right away dor DCE and which ones we could assume can be done externally (e.g. registration, brain masking, etc) would be very valuable.

**Comment by b3nbm** [comment]
Just to notice that we are going to contribute on perfusion models during the MRathon. I'll focus on adding a simple concentation calculation and extended Tofts pharmacokinetics model that will be validated using the QIBA_v4 DCE digital reference object commonly used in the field. More concentration equations and pharamacokinetics models should be added on longer term.

**Comment by slevyrosetti** [comment]
- Example DSC data can be added from Dynamic Susceptibility Contrast MRI toolbox (https://github.com/marcocastellaro/dsc-mri-toolbox), see demo-data.
- the filename in Perfusion model directory (`DCE.m`) should be `DSC.m`

**Comment by b3nbm** [comment]
I'm trying to make voxelwise DCE fitting to work. 

I am using the preprocessing method function PrecomputeData to compute the AIF over the 4D dataset from a mask before voxelwise fittin. However this function does not seem to be called by the "Voxel fit" option, hence I cannot compute AIF.

Is there a way to force a 4D preprocessing step (AIF from mask) prior to 1D fitting?

**Comment by agahkarakuzu** [comment]
@b3nbm can you make sure that you have defined the method as `Precompute`/`PrecomputeData` but not as `PreCompute` ?

**Comment by b3nbm** [comment]
the method is defined as `PrecomputeData`

**Comment by notZaki** [comment]
@b3nbm 
I tried your branch and it looks like `PrecomputeData` does get called.

I just used junk data for a quick test, using the below code:
```
clear
Model = dce;
[nx, ny, nz, nt] = deal(2,1,1,661);
data.DCEData = randn(nx, ny, nz, nt) + 5;
data.T1Map = randn(nx, ny, nz) + 5;
data.MaskAIF = zeros(size(data.T1Map));
data.MaskAIF(1,1,1) = 1;
fitObj = FitData(data, Model);
plot(fitObj.Model.aif)
```
and I can see an AIF. 
It's not a pretty AIF, but that's not the code's fault.
